go/vt/mysqlctl/mysqlshellbackupengine.go

```diff
@@ -0,0 +1,340 @@
+package mysqlctl
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os/exec"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/spf13/pflag"
+
+	"vitess.io/vitess/go/mysql/replication"
+	"vitess.io/vitess/go/vt/mysqlctl/backupstorage"
+	tabletmanagerdatapb "vitess.io/vitess/go/vt/proto/tabletmanagerdata"
+	"vitess.io/vitess/go/vt/servenv"
+	"vitess.io/vitess/go/vt/vterrors"
+)
+
+var (
+	// location to store the mysql shell backup
+	mysqlShellBackupLocation = ""
+	// flags passed to the mysql shell utility, used both on dump/restore
+	mysqlShellFlags = "--defaults-file=/dev/null --js -h localhost"
+	// flags passed to the Dump command, as a JSON string
+	mysqlShellDumpFlags = `{"threads": 2}`
```

Comment:
I would say because MySQL Shell backups can be a bit more intensive - we are requesting a bunch of data off MySQL which needs to be fetched, parsed and compressed - and in our particular use case we were taking backups online so we didn't wan't it to cause that much disruption. It is also part of the reason why I made sure  `ShouldDrainForBackup()` was configurable in case it is more suitable for the use case.

I am fine with changing the default to `runtime.NumCPU()` though since it is configurable and leave this up to the user to decide based on their environment requirements, although I am also conscious that it might cause some issues in Kube where it will show the number of CPUs of the underlying node despite the pod being possibly limited on how much CPU it can use.

Reply:
Seems like we already have a `--concurrency` flag for the `Backup` command. Should we reuse that flag?

Reply:
so looking this up again today, the default in MySQL Shell if not passed is `4`. what I have been considering is that if we do something like:
```go
mysqlShellDumpFlags = fmt.Sprintf("{\"threads\": %d}", runtime.NumCPU())
```

it might give the user the wrong impression that this is the default for MySQL Shell and that if they use the flag to change other settings and they don't add `thread` to it, they will use all CPUs by default. 

so perhaps we should just set it to `{"threads": 4}` instead?

Reply:
> it might give the user the wrong impression that this is the default for MySQL Shell and that if they use the flag to change other settings and they don't add thread to it, they will use all CPUs by default.

This is a documentation concern. We can't expect any user expectation. I can say that I would not assume as much.

Reply:
still, we can try to make it more difficult for users to do the wrong thing. perhaps we can remove `threads` from the default value and leave it up to the user to modify it if the desire so?